File: src/OpenPlexus/Worlds/Snake.cs
namespace OpenPlexus.Worlds;

/// <summary>What a cell in the view contains.</summary>
/// <remarks>
/// <b>Categorical, and quantised one-hot rather than by hyperplane.</b> These
/// are 0 1 2 3 and a hyperplane over those numbers would make wall-and-body
/// near and empty-and-food far, which is arithmetic nobody meant.
/// </remarks>
public enum Cell { Empty, Wall, Body, Food }

/// <summary>Where the snake goes next.</summary>
public enum SnakeAction { North, South, East, West }

/// <summary>One cell of the view, offset from the head.</summary>
public readonly record struct Seen(int Dx, int Dy, Cell Content);

/// <summary>
/// What the snake can see. <b>Head-centred and local.</b>
/// </summary>
/// <remarks>
/// <b>Centred</b> so the same situation in two places is one observation, which
/// is what makes anything recur at all. <b>Local</b> so the food is usually
/// unseen, which is what gives <i>act to disambiguate</i> something to
/// disambiguate.
/// </remarks>
public sealed record SnakeView
{
    public required IReadOnlyList<Seen> Cells { get; init; }
}

/// <summary>
/// How the world is set up. Every constant named and none defaulted.
/// </summary>
/// <remarks>
/// <b>A constant that never changes looks like the background.</b> Requiring
/// each one is how a number gets set on purpose rather than inherited.
/// </remarks>
public sealed record SnakeSettings
{
    public required int Width { get; init; }
    public required int Height { get; init; }

    /// <summary>
    /// How far the snake sees, as a radius. <b><c>null</c> is the whole board,
    /// still centred</b> — the two are arms of one experiment, not a feature
    /// and its disabled state.
    /// </summary>
    public int? Sight { get; init; }

    public required double StartingEnergy { get; init; }
    public required double EnergyPerStep { get; init; }
    public required double EnergyPerFood { get; init; }
}

/// <summary>
/// An interactive environment, not a recorded corpus.
/// </summary>
/// <remarks>
/// The only source supplying an error signal for free. Chosen over ARC-AGI-3 to
/// start with, so that a bad result cannot be blamed on the environment.
/// </remarks>
public sealed class Snake
{
    private readonly record struct Point(int X, int Y);

    private readonly SnakeSettings _settings;
    private readonly Random _rng;

    /// <summary>Head first. The tail is the last entry.</summary>
    private readonly List<Point> _body = [];

    private readonly HashSet<Point> _occupied = [];

    private Point _food;
    private double _energy;
    private bool _alive = true;

    public Snake(SnakeSettings settings, int seed)
    {
        ArgumentNullException.ThrowIfNull(settings);
        ArgumentOutOfRangeException.ThrowIfLessThan(settings.Width, 5);
        ArgumentOutOfRangeException.ThrowIfLessThan(settings.Height, 5);
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(settings.StartingEnergy);

        _settings = settings;
        _rng = new Random(seed);
        _energy = settings.StartingEnergy;

        // Placed at the centre rather than at random: a start position nobody
        // chose is a constant that would sit under every number.
        var head = new Point(settings.Width / 2, settings.Height / 2);
        foreach (var offset in (int[])[0, 1, 2])
        {
            var part = head with { X = head.X - offset };
            _body.Add(part);
            _occupied.Add(part);
        }

        _food = PlaceFood();
    }

    /// <summary>Whether the run is still going.</summary>
    public bool Alive => _alive;

    /// <summary>
    /// Something to lose.
    /// </summary>
    /// <remarks>
    /// Depletes, food restores it, running out <b>ends</b> the run rather than
    /// resetting. <b>Nothing declares food good.</b> A policy that does not eat
    /// gets fewer steps of experience — selection without a reward, and the
    /// first source of preference this design has had.
    /// </remarks>
    public double Energy => _energy;

    /// <summary>How long the snake is. Grows by one per fruit.</summary>
    public int Length => _body.Count;

    /// <summary>Advances one tick.</summary>
    /// <remarks>
    /// <b>A dead run does not step.</b> Running out of energy ends it; nothing
    /// here resets, because a run that restarts on failure has nothing at stake.
    /// </remarks>
    public void Step(SnakeAction action)
    {
        if (!_alive) throw new InvalidOperationException("the run is over");

        var (dx, dy) = action switch
        {
            SnakeAction.North => (0, -1),
            SnakeAction.South => (0, 1),
            SnakeAction.East => (1, 0),
            SnakeAction.West => (-1, 0),
            _ => throw new ArgumentOutOfRangeException(nameof(action)),
        };

        var head = _body[0];
        var next = new Point(head.X + dx, head.Y + dy);

        if (OutOfBounds(next) || _occupied.Contains(next))
        {
            _alive = false;
            return;
        }

        _energy -= _settings.EnergyPerStep;

        var ate = next == _food;
        _body.Insert(0, next);
        _occupied.Add(next);

        if (ate)
        {
            _energy += _settings.EnergyPerFood;
            _food = PlaceFood();
        }
        else
        {
            var tail = _body[^1];
            _body.RemoveAt(_body.Count - 1);
            _occupied.Remove(tail);
        }

        if (_energy <= 0.0) _alive = false;
    }

    /// <inheritdoc cref="SnakeView"/>
    public SnakeView View()
    {
        var head = _body[0];
        var cells = new List<Seen>();

        if (_settings.Sight is int sight)
        {
            for (var dy = -sight; dy <= sight; dy++)
                for (var dx = -sight; dx <= sight; dx++)
                    cells.Add(new Seen(dx, dy, At(head.X + dx, head.Y + dy)));
        }
        else
        {
            // Still centred. The whole board and a local window are arms of one
            // experiment, so they must produce the same KIND of thing.
            for (var y = 0; y < _settings.Height; y++)
                for (var x = 0; x < _settings.Width; x++)
                    cells.Add(new Seen(x - head.X, y - head.Y, At(x, y)));
        }

        return new SnakeView { Cells = cells };
    }

    private Cell At(int x, int y)
    {
        var point = new Point(x, y);
        if (OutOfBounds(point)) return Cell.Wall;
        if (_occupied.Contains(point)) return Cell.Body;
        return point == _food ? Cell.Food : Cell.Empty;
    }

    private bool OutOfBounds(Point point) =>
        point.X < 0 || point.Y < 0 ||
        point.X >= _settings.Width || point.Y >= _settings.Height;

    private Point PlaceFood()
    {
        var free = _settings.Width * _settings.Height - _occupied.Count;
        if (free <= 0) throw new InvalidOperationException("the board is full");

        while (true)
        {
            var candidate = new Point(_rng.Next(_settings.Width), _rng.Next(_settings.Height));
            if (!_occupied.Contains(candidate)) return candidate;
        }
    }
}

File: openplexus/tasks/snake.py
"""A world that answers back. Snake, with a local view so acting is informative.

John's suggestion, 2026-08-01, as the step before ARC-AGI-3. It supplies the two
things a recorded corpus cannot:

- **An action changes what is observed.** Every mechanism this project still
  lacks — prediction, the trigger for asking, an output channel — needs that,
  and no corpus can say whether acting to disambiguate beats watching.
- **Recurrence.** The same local configuration happens thousands of times, which
  is what counting needs before any statistic exists. ARC withholds recurrence by
  design; this gives it away.

## Why the view is LOCAL, and it is the whole point

Full-board snake is deterministic and fully observed, so everything worth
knowing is already on screen and a system that learns it has learned dynamics
rather than concepts. That is worth having and it does not test the claim.

With a window of `sight` cells, the food is usually off-view, so **moving is how
you find out where it is**. That gives "act to disambiguate" something to
disambiguate, and it gives the watching baseline a precise meaning: a passive
observer of somebody else's game, at the same number of observations.

`sight=None` is the full board, so the two are arms of one experiment rather
than two programs.

## What it is deterministic for

The dynamics have **no noise at all**. That is deliberate and it is the reason
to run this before anything stochastic: uncertainty sampling's known failure is
chasing irreducible noise, and here there is none, so a prediction-error-driven
policy cannot fail that way by construction. Adding stochasticity later is how
that failure mode gets tested, separately, rather than confounded into the first
measurement.

## What prediction needs, and it is already built

Predicting the next observation from the current one AND the action is a TRIPLE,
not a pair, so `CoOccurrence` cannot hold it — and `openplexus/composition.py`
already counts exactly this shape: `Composition(left, right, target)` with
`left` the current code, `right` the action and `target` the next code. No new
mechanism, and it is the same class the typed walk uses for route kinds.

## Conventions

Dependency-free, like every other task here: the model layer may use numpy and
the thing a result is measured against may not.
"""

from __future__ import annotations

import random
from dataclasses import dataclass

#: The four things the machine can do. Absolute rather than relative, because
#: John's framing is buttons on a device and a button does not know which way
#: the snake is facing.
ACTIONS = ((0, -1), (0, 1), (-1, 0), (1, 0))

#: What a cell can contain. These are the alphabet an observation is written in,
#: and they are deliberately few: a front end that has to tell four things apart
#: is doing quantisation, not perception.
EMPTY, WALL, BODY, FOOD = 0, 1, 2, 3


@dataclass
class Step:
    """One transition, and everything needed to score a prediction of it.

    Attributes:
        view: What was observable AFTER the action, as flat cell values.
        action: Which action was taken to get here.
        ate: Whether the food was taken on this step. **Not a reward** — nothing
            here optimises it; it is reported so an experiment can say whether
            a policy that never sees a reward nonetheless eats more.
        died: Whether the snake ran into a wall or itself.
    """

    view: tuple[int, ...]
    action: int
    ate: bool
    died: bool


class Snake:
    """The board, and what one machine can see of it.

    Attributes:
        width, height: Board size.
        sight: Half-width of the visible window, or `None` for the whole board.
            A window of `sight=2` shows a 5x5 patch centred on the head.
    """

    def __init__(self, width: int = 12, height: int = 12,
                 sight: int | None = 2, seed: int = 0) -> None:
        if width < 3 or height < 3:
            raise ValueError("a board smaller than 3x3 has no room to turn")
        if sight is not None and sight < 1:
            raise ValueError("a sight of zero sees only the head, which never "
                             "changes, so no action could ever be informative")
        self.width, self.height, self.sight = width, height, sight
        self._rng = random.Random(seed)
        self.reset()

    def reset(self) -> None:
        self.body = [(self.width // 2, self.height // 2)]
        self.heading = 3
        self.dead = False
        self._place_food()

    def _place_food(self) -> None:
        free = [(x, y) for x in range(self.width) for y in range(self.height)
                if (x, y) not in self.body]
        self.food = self._rng.choice(free) if free else None

    def _cell(self, x: int, y: int) -> int:
        if not (0 <= x < self.width and 0 <= y < self.height):
            return WALL
        if (x, y) in self.body:
            return BODY
        return FOOD if (x, y) == self.food else EMPTY

    def view(self) -> tuple[int, ...]:
        """What is observable now, flattened.

        **Centred on the head**, so the same situation in two places on the
        board produces the same observation. That is what makes the recurrence
        this task exists to supply actually recur.
        """
        if self.sight is None:
            return tuple(self._cell(x, y)
                         for y in range(self.height)
                         for x in range(self.width))
        head_x, head_y = self.body[0]
        span = range(-self.sight, self.sight + 1)
        return tuple(self._cell(head_x + dx, head_y + dy)
                     for dy in span for dx in span)

    def step(self, action: int) -> Step:
        """Take one action and return what followed.

        A death resets the board and is reported rather than raised: the stream
        never ends, which is what C4 asks of everything here.
        """
        if not 0 <= action < len(ACTIONS):
            raise ValueError(f"no action {action}; there are {len(ACTIONS)}")
        dx, dy = ACTIONS[action]
        head_x, head_y = self.body[0]
        ahead = (head_x + dx, head_y + dy)
        died = (not (0 <= ahead[0] < self.width
                     and 0 <= ahead[1] < self.height)
                or ahead in self.body[:-1])
        ate = False
        if died:
            self.reset()
        else:
            self.body.insert(0, ahead)
            ate = ahead == self.food
            if ate:
                self._place_food()
            else:
                self.body.pop()
            self.heading = action
        return Step(view=self.view(), action=action, ate=ate, died=died)
